lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm

```diff
@@ -205,6 +205,203 @@ sub schedules {
+                my $found = 0;
+                foreach my $test_suite (@{$yaml->{$group}->{architectures}->{$arch}->{$product}}) {
+                    if (ref $test_suite eq ref {}) {
+                        foreach my $name (keys %$test_suite) {
+                            next unless $name eq $template->test_suite->name;
+                            $found = 1;
+
+                            my $attr = $test_suite->{$name};
+                            if (%$attr) {
+                                if ($attr->{machine}) {
+                                    my $machine
+                                      = $self->schema->resultset("Machines")->find({name => $attr->{machine}});
+                                    if (!$machine) {
+                                        push @$errors, "Machine '" . $attr->{machine} . "' invalid";
+                                        next;
+                                    }
+                                    $template->machine_id($machine->id);
+                                }
+                                if ($attr->{priority}) {
+                                    $template->prio($attr->{priority});
+                                }
+                                eval { $template->update; };
+                                push @$errors, $@ if $@;
+                            }
+                        }
+                    }
+                    elsif ($test_suite eq $template->test_suite->name) {
+                        $found = 1;
+
+                        # Apply defaults
+                        my $machine_name = $yaml->{$group}->{defaults}->{$arch}->{machine};
+                        my $machine      = $self->schema->resultset("Machines")->find({name => $machine_name});
+                        push @$errors, "Machine '" . $machine_name . "' invalid" unless $machine;
+                        eval {
+                            $template->update(
+                                {
+                                    prio       => $yaml->{$group}->{defaults}->{$arch}->{priority},
+                                    machine_id => $machine->id,
+                                });
+                        };
+                        push @$errors, $@ if $@;
+                    }
+                }
+
+                next if $found;
+                # Drop the existing test suite which is not specified
+                eval { $template->delete; };
+                push @$errors, $@ if $@;
+            }
+        }
+
+        foreach my $arch (keys %{$yaml->{$group}->{architectures}}) {
+            foreach my $product_name (keys %{$yaml->{$group}->{architectures}->{$arch}}) {
+                foreach my $spec (@{$yaml->{$group}->{architectures}->{$arch}->{$product_name}}) {
+
+                    my $testsuite_name;
+                    my $prio;
+                    my $machine_name;
+                    if (ref $spec eq ref {}) {
```

Comment:
Yes, `ref $spec eq 'HASH'` is correct. And there should be a small performance difference. Since `{}` needs to be allocated as a new hash every time, while `'HASH'` is optimized as a constant.